File: 卡码/代码随想录/1/11-图论/C++/graph/dfs/LeetCode_0827.cpp
//827. 最大人工岛
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
	int dir[4][2]={{0,1},{1,0},{0,-1},{-1,0}};
	int cnt=0;
public:
	void dfs(vector<vector<int>>& grid,int x,int y,int mark){
		grid[x][y]=mark;
		cnt++;
		for(int i=0;i<4;i++){
			int nextX=x+dir[i][0];
			int nextY=y+dir[i][1];
			if(nextX<0||nextX>grid.size()-1||nextY<0||nextY>grid[0].size()-1) continue;
			if(grid[nextX][nextY]==0||grid[nextX][nextY]>1) continue;
			dfs(grid,nextX,nextY,mark);
		}
	}
	int largestIsland(vector<vector<int>>& grid) {
		int n=grid.size(),m=grid[0].size();
		unordered_map<int,int> areas;
		int mark=2;
		bool isAllLand=true;
		for(int i=0;i<n;i++){
			for(int j=0;j<m;j++){
				if(grid[i][j]==0) isAllLand=false;
				if(grid[i][j]==1){
					cnt=0;
					dfs(grid,i,j,mark);
					areas[mark]=cnt;
					mark++;
				}
			}
		}
		if(isAllLand) return n*m;
		int res=0;
		unordered_set<int> visited;
		for(int i=0;i<n;i++){
			for(int j=0;j<m;j++){
				int count=1;
				visited.clear();
				if(grid[i][j]==0){
					for(int k=0;k<4;k++){
						int x=i+dir[k][0];
						int y=j+dir[k][1];
						if(x<0||x>n-1||y<0||y>m-1) continue;
						if(grid[x][y]>1){
							if(visited.count(grid[x][y])) continue;
							count+=areas[grid[x][y]];
							visited.insert(grid[x][y]);
						}
					}
				}
				res=max(res,count);
			}
		}
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/Python/图论/dfs/LeetCode_0827.py
#827. 最大人工岛
from typing import *

class Solution:
  def __init__(self):
    self.dir:List[List[int]]=[[0,1],[1,0],[0,-1],[-1,0]]
    self.cnt:int = 0
  def dfs(self, grid: List[List[int]], x: int, y: int, mark: int):
    self.cnt += 1
    grid[x][y] = mark
    for i in range(4):
      nextX, nextY = x + self.dir[i][0], y + self.dir[i][1]
      if nextX < 0 or nextX > len(grid) - 1 or nextY < 0 or nextY > len(grid[0]) - 1:
        continue
      if grid[nextX][nextY] == 0 or grid[nextX][nextY] > 1:
        continue
      self.dfs(grid, nextX, nextY, mark)
  def largestIsland(self, grid: List[List[int]]) -> int:
    n, m = len(grid), len(grid[0])
    areas: dict[int, int] = dict()
    isAllLand: bool = True
    mark: int = 2
    for i in range(n):
      for j in range(m):
        if grid[i][j] == 0:
          isAllLand = False
        if grid[i][j] == 1:
          self.cnt = 0
          self.dfs(grid, i, j, mark)
          areas[mark] = self.cnt
          mark += 1
    if isAllLand:
      return n * m
    visited: set[int] = set()
    res: int = 0
    for i in range(n):
      for j in range(m):
        if grid[i][j] == 0:
          count: int = 1
          visited.clear()
          for k in range(4):
            x, y = i + self.dir[k][0], j + self.dir[k][1]
            if x < 0 or x > n - 1 or y < 0 or y > m - 1:
              continue
            if grid[x][y] in visited:
              continue
            visited.add(grid[x][y])
            count += areas.get(grid[x][y], 0)
          res = max(res, count)
    return res